### ISSUE TITLE
Any options to disable TLS v1.0 and TLS v1.1 for the cluster-agent?

### ISSUE THREAD

**Issue by thaviphone** [canonical/microk8s-cluster-agent#12]
Hi,

we are using Microk8s for quite some time now.  Our lastest company's Qualys-Scan revealed a vulnerability under port 25000:
# microk8s version
MicroK8s v1.26.0 revision 4390

Trying to configure the tags as decribed in the staled issue: https://github.com/canonical/microk8s/issues/2896 did not have the desired result.  The cluster-agent seem to refuse to start using these tags:

e.g. --ssl-version or --cipher
Jan 24 17:21:09 *** systemd[1]: Started Service for snap application microk8s.daemon-cluster-agent.
Jan 24 17:21:09 *** microk8s.daemon-cluster-agent[3488952]: Error: unknown flag: --ssl-version

Are there other than that other options to disable TLS v1.0 and TLS v. 1.1 ?

**Comment by neoaggelos** [comment]
Hi @thaviphone, thank you for reporting this

Indeed, after testing I also found some issues with cluster-agent accepting TLS v1.0 and v1.1 connections. I'll investigate further and get back to you.

**Comment by KarthikK123** [comment]
Hi @neoaggelos , is there any update on this?

**Comment by neoaggelos** [comment]
Hi @KarthikK123

Apologies, haven't had time to look into this further just yet. It is a bit puzzling to me because starting a simple server with `http.ListenAndServeTLS()` does not accept 1.0, 1.1 TLS connections, but this is not the case for the cluster agent. I believe we have probably missed something obvious here.

**Comment by neoaggelos** [comment]
Hi @thaviphone @KarthikK123. Following up after some investigation. With the latest MicroK8s 1.26/stable, I am getting the following:

```
for ver in tls1 tls1_1 tls1_2 tls1_3; do openssl s_client -connect localhost:25000 -$ver </dev/null > /dev/null 2>/dev/null && echo $ver worked || echo $ver failed; done
tls1 failed
tls1_1 failed
tls1_2 worked
tls1_3 worked
```

Can you check whether the cluster-agent accepts TLSv1.0 and TLSv1.1 connections on your end? Thanks!

For more context: this was an issue in the past (pre 1.24) where the cluster-agent was served using an old version of gunicorn. Since 1.24, the cluster-agent is a Go service and this should not be an issue any more.

**Comment by KarthikK123** [comment]
I just tried this out, I am still seeing v1.0 and v1.1 are being accepted.

![image](https://user-images.githubusercontent.com/49305872/225831723-6c92146c-4eea-4dd5-bbf5-7e4998aa9804.png)


I'm on microk8s v1.26.1, revision 4595

**Comment by thaviphone** [comment]
confirming having the same on my end...  output is excactly the same...
I'm also on microk8s v1.26.1 revision 4595

**Comment by thaviphone** [comment]
by far I'm not a programer/coder or anything...  I know nothing about Go...  I'm not even able to read your code...  not to mention on the Go service, which is used...  but I can think of the Go service being an "web server service"...  so in my simplest understanding and thoughts of a web service, there must be an option to set a kind of min tls version to be used...  is there possibly something missing?  Maybe an option in the Go service setting a min version of tls?

**Comment by neoaggelos** [comment]
Yes, setting the MinTLSVersion should suffice. I am however unable to understand why this is different in the first place. Might I ask which operating system you are using, or if there is anything different in your setup?

```
ubuntu@u1:~$ sudo snap install microk8s --classic --channel 1.26
microk8s (1.26/stable) v1.26.1 from Canonical✓ installed
ubuntu@u1:~$ for ver in tls1 tls1_1 tls1_2 tls1_3; do openssl s_client -connect localhost:25000 -$ver </dev/null > /dev/null 2>/dev/null && echo $ver worked || echo $ver failed; done
tls1 failed
tls1_1 failed
tls1_2 worked
tls1_3 worked
ubuntu@u1:~$ sudo snap list microk8s
Name      Version  Rev   Tracking     Publisher   Notes
microk8s  v1.26.1  4595  1.26/stable  canonical✓  classic
```

I'm getting this when I try on a fresh focal VM

**Comment by ghost** [comment]
I’ve been getting TLS1 and 1.1 on fresh installs of Ubuntu 22.04 and 22.10. Installation of Microk8s though Snap as usual then setting up 3 host HA. 

Verified on Nessus scans.

**Comment by KarthikK123** [comment]
I tested on a fresh Azure VM running Ubuntu 20.04.

**Comment by thaviphone** [comment]
I'm running a four node cluster on Rocky Linux 8.0

**Comment by KarthikK123** [comment]
Hi @neoaggelos, did you get a chance to investigate further? This is still being flagged in our company scans

**Comment by neoaggelos** [comment]
Hi @KarthikK123 

Okay, finally got it, `nmap --script ssl-enum-ciphers` to the rescue. We will add an option to set the min TLS version accepted. I am also keen on having it set to TLS1.2 by default starting from 1.27 (I'm not as excited to change the defaults for microk8s 1.25 or 1.26 at the moment, out of fear of breaking existing environments). For older versions, I expect setting something like `--min-tls-version=tls12` would be an acceptable tradeoff

**Comment by neoaggelos** [comment]
Hi @KarthikK123 @thaviphone and anyone else that is interested in this. The linked PR is merged and there is a new build out on `1.26/edge` and `1.25/edge` channels. No extra arguments are needed. The build will be pushed to `1.26/stable` and `1.25/stable` respectively with the next patchlevel release, expected in a few days.

```
sudo snap refresh microk8s --channel 1.26/edge
```

**Comment by KarthikK123** [comment]
Thanks for the update @neoaggelos! Please let me know once the change is pushed to the stable channel so that I can verify that it is not being flagged on our end anymore.
